tls/s2n_cipher_suites.c

```diff
@@ -1168,27 +1171,65 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
     return 0;
 }
 
+bool s2n_cipher_suite_uses_chacha20_alg(struct s2n_cipher_suite *cipher_suite) {
+    return cipher_suite && cipher_suite->record_alg && cipher_suite->record_alg->cipher == &s2n_chacha20_poly1305;
+}
+
+/* Iff the server has enabled allow_chacha20_boosting and the client has a chacha20 cipher suite as its most 
+ * preferred cipher suite, then we have mutual chacha20 boosting support.
+ */
+static S2N_RESULT s2n_mutual_chacha20_boosting_support(const struct s2n_cipher_preferences *cipher_preferences, struct s2n_stuffer wire_stuffer, uint32_t cipher_suite_len, bool *mutual_chacha20_boosting_support) {
+    RESULT_ENSURE_REF(mutual_chacha20_boosting_support);
+    RESULT_ENSURE_REF(cipher_preferences);
+    *mutual_chacha20_boosting_support = false;
+
+    if (!cipher_preferences->allow_chacha20_boosting) {
+        return S2N_RESULT_OK;
+    }
+
+    uint8_t clients_first_cipher_iana[S2N_SSLv2_CIPHER_SUITE_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&wire_stuffer, clients_first_cipher_iana, cipher_suite_len));
+
+    struct s2n_cipher_suite *client_first_cipher_suite = NULL;
+    RESULT_GUARD(s2n_cipher_suite_from_iana(clients_first_cipher_iana, &client_first_cipher_suite));
+    RESULT_ENSURE_REF(client_first_cipher_suite);
+
+    if (s2n_cipher_suite_uses_chacha20_alg(client_first_cipher_suite)) {
+        *mutual_chacha20_boosting_support = true;
+    }
+
+    return S2N_RESULT_OK;
+}
+
 static int s2n_set_cipher_as_server(struct s2n_connection *conn, uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
 {
     POSIX_ENSURE_REF(conn);
     POSIX_ENSURE_REF(conn->secure);
 
+    struct s2n_blob wire_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&wire_blob, wire + cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN, count * cipher_suite_len));
```

Comment:
This is (almost)  equivalent to the original implementation with one little exception. All the tests are using setting as a tls server but I can add another one for setting as sslv2 server.

My understanding of sslv2 cipher suites is that it's encoded with 3 bytes instead of >= sslv3's 2 bytes. For sslv2, I think the first byte encodes the length of the key. In order to have sslv2-compatible cipher suites in sslv3 and beyond, many people just zero or entirely ignore the first byte (openssl does this [here](https://github.com/openssl/openssl/blob/3a09dfb4f9aace93d2c20d6d1b4968cc583884d6/ssl/ssl_lib.c#L6606)). So yes, we are essentially skipping the first byte for sslv2 ciphers and we only consider the last two remaining bytes for the IANA. 

This implementation does the same as the original:

1. If !sslv2 then the very first byte of the wire is a cipher suite. Else, skip by one byte.
2. Iterate by `cipher_suite_len` bytes to retrieve the nth cipher suite (regardless if sslv2 or not)
3. Only read `S2N_TLS_CIPHER_SUITE_LEN` or 2 bytes of pointer.

This worked with the previous code and my current change because we discard the leading zero and read only two bytes. 

Tried illustrating this below:

```
// Example: sslv2 client wire

retrieving the first iana:

0x00 | 0x01 0x02 0x00 | 0x03 0x04 0x00 0x05 0x06

^ We skipped the first byte (sslv2) and read `cipher_suite_len = 3` bytes. To retrieve the actual iana, we compare the first `S2N_TLS_CIPHER_SUITE_LEN = 2` bytes (0x01 0x02).

retrieving the second iana:

0x00 0x01 0x02 0x00 | 0x03 0x04 0x00 | 0x05 0x06

^ Same as before. We read `cipher_suite_len = 3` bytes and compare the first two bytes (0x03 0x04).

retrieving the third and final iana:

0x00 0x01 0x02 0x00 0x03 0x04 0x00 | 0x05 0x06 ??? |

^ this is where it gets weird. We still read `cipher_suite_len = 3` bytes, which means the last byte we read in is undefined (since there is only two bytes left). 
However, we still only compare the first two bytes (0x05 0x06) and this is defined. 
```

So this still does the same thing as before except for the last iteration of an sslv2 server, it reads in 2 defined bytes + one undefined one. However, this code only ever reads the first 2 defined bytes so we get the desired behaviour.

I can see how this is can bother people though. The problem with the stuffer implementation is that we to need load either two or three bytes into memory before comparing. Because we don't know, we will always need to initialise `S2N_SSLv2_CIPHER_SUITE_LEN` bytes since we may need all 3 bytes. Therefore, if we are not negotiating sslv2 then we will have an extra byte in memory undefined (but it will never get used). The original implementation did not face this problem because it just performed pointer arithmetic and never needed to load any more info in memory.

Reply:
You can't read past the end of a stuffer. That will cause an error. Even if you could, you should NEVER be reading past the end of defined memory. That's incredibly dangerous, and the main reason blobs and stuffers exist.

If there was a test for this that read to the end, that test would have failed. That test needs to exist if we're going to touch this code.

You either need to figure out how to do a valid memory read with the stuffer, or restore the original pointer arithmetic.

Reply:
Oh no, it WON'T give you an error, because you're lying about the length of the data. You tell it count * cipher_suite_len.  You really are reading invalid memory. You need to set your blob up accurately-- that is a recipe for disaster.